client/src/components/Minimap/Minimap.tsx

```diff
@@ -0,0 +1,238 @@
+  const observer = useRef<IntersectionObserver | null>(null);
+  const lastIntersections = useRef({});
+  const updateLinks = useRef<CallableFunction | null>(null);
+  const listRef = useRef<HTMLOListElement>(null);
+
+  // Keep track of all the different ways the minimap can be opened and closed.
+  const onMouseOver = () => {
+    if (window.matchMedia('(hover: hover)').matches) {
+      // Opening with hover should not keep the menu expanded.
+      setExpanded(true);
+    }
+  };
+  const onMouseOut = () => {
+    if (window.matchMedia('(hover: hover)').matches) {
+      if (!keepExpanded) {
+        setExpanded(false);
+        setKeepExpanded(false);
+      }
+    }
+  };
+  const onClickToggle = () => {
+    setExpanded(!expanded);
+    setKeepExpanded(!expanded);
+  };
+  const onClickLink = (e: React.MouseEvent) => {
+    // Prevent navigating if the link is only partially shown.
+    if (!expanded) {
+      e.preventDefault();
+    }
+    setExpanded(true);
+    setKeepExpanded(true);
+  };
+
+  useEffect(() => {
+    const onClickOutside = (e: MouseEvent) => {
+      if (container.contains(e.target as HTMLElement)) {
+        return;
+      }
+      setExpanded(false);
+      setKeepExpanded(false);
+    };
+    document.addEventListener('click', onClickOutside, true);
+  }, []);
+
+  /**
+   * Performance-sensitive intersections calculations with a double debounce:
+   * - With the IntersectionObserver API, the browser decides how often to update us, compared to constant `scroll`.
+   * - We keep track of intersecting elements on every IntersectionObserver update,
+   * - but only update the links after updates have stopped for 100ms.
+   */
+  useEffect(() => {
+    const obsCallback = (newEntries) => {
+      lastIntersections.current = newEntries.reduce(mapIntersections, {
+        ...lastIntersections.current,
+      });
+
+      if (!updateLinks.current) {
+        updateLinks.current = debounce((newIntersections) => {
+          setIntersections(newIntersections);
+          updateScrollPosition(listRef.current as HTMLOListElement);
+        }, 100);
+      }
+
+      updateLinks.current(lastIntersections.current);
+
+      // Support for InlinePanel removals: when they stop intersecting, re-render the whole minimap.
+      newEntries.forEach(({ target }) => {
+        const deletedInlinePanel = target.closest('.deleted');
+        if (deletedInlinePanel) {
+          onUpdate(container);
+        }
+      });
+    };
+
+    if (!observer.current) {
+      observer.current = new IntersectionObserver(obsCallback, observerOptions);
+    }
+
+    const obs = observer.current as IntersectionObserver;
+
+    obs.disconnect();
+
+    links.forEach(({ panel, toggle }) => {
+      // Special-case for top-level InlinePanel and StreamField, where the
+      // link only shows as active if the anchor is in view.
+      const isTopLevelNested =
+        panel.matches('.w-panel--nested') &&
+        panel.closest('[data-field]') === null;
```

Comment:
I’m a bit annoyed we have to do this but otherwise it looks a bit odd when navigating StreamField.